src/tagmanager/tm_query.c

```diff
@@ -0,0 +1,314 @@
+ *
+ * The results are generally sorted by tag name, even if @a sort_attr is NULL.
+ * supplied. The results can be
+ * sorted by @a sort_attr, and deduplicated by @a dedup_attr. Deduplication
+ * happens after sorting. There is an additional limitation for deduplication:
+ * Only adjacent tags can be deduplicated so they must be suitably ordered.
+ * This can be achieved by using the same attributes for @a sort_attr and
+ * @a dedup_attr.
+ *
+ * The results are stored in a @c GPtrArray. The individual tags are still
+ * owned by the workspace but the container array must be freed with
+ * g_ptr_array_free().
+ *
+ * @param q The query to execute
+ * @param sort_attr @nullable @array{zero-terminated=1} Tag attribute to use
+ *                   as sort key, or @c NULL to not sort.
+ * @param dedup_attr @nullable @array{zero-terminated=1} Tag attribute to use
+ *                    as deduplicaiton key, or @c NULL to not perform deduplicaiton.
+ * @return @transfer{container} @elementtype{TMTag} Resulting tag array.
+ */
+GEANY_API_SYMBOL
+GPtrArray *tm_query_exec(TMQuery *q, TMTagAttrType *sort_attr, TMTagAttrType *dedup_attr)
+{
+	TMTag **tags, *tag;
+	gint ntags;
+	guint i;
+	GString *s;
+	GQueue res = G_QUEUE_INIT;
+	TMSortOptions sort_options;
+	GList *node, *next;
+	gchar *scope;
+	TMParserType *lang;
+	GPtrArray *ret;
+
+	sort_options.sort_attrs = NULL;
+	/* tags_array isn not needed by tm_tag_compare(), but for tm_search_cmp() */
+	sort_options.tags_array = NULL;
+	sort_options.first = TRUE;
+
+	foreach_ptr_array(s, i, q->names)
+	{
+		TMTag **ptag;
+		sort_options.cmp_len = s->len;
+		if (q->data_sources & TM_QUERY_SOURCE_GLOBAL_TAGS)
+		{
+			tags = tm_tags_find(q->workspace->global_tags, s->str, s->len, &ntags);
+			foreach_c_array(ptag, tags, ntags)
+				g_queue_insert_sorted(&res, *ptag, tag_compare_ptr, &sort_options);
+		}
+		if (q->data_sources & TM_QUERY_SOURCE_SESSION_TAGS)
+		{
+			tags = tm_tags_find(q->workspace->tags_array, s->str, s->len, &ntags);
+			foreach_c_array(ptag, tags, ntags)
+				g_queue_insert_sorted(&res, *ptag, tag_compare_ptr, &sort_options);
+		}
+	}
+
+	/* Filter tags according to scope, type and lang. */
+	for (node = res.head; node; node = next)
+	{
+		gboolean match = TRUE;
+
+		next = node->next;
+		tag = node->data;
+		foreach_ptr_array(scope, i, q->scopes)
+			match = match && (g_strcmp0(tag->scope, scope) == 0);
+
+		foreach_array(gint, lang, q->langs)
+			match = match && tm_tag_langs_compatible(*lang, tag->lang);
+
+		if (match && q->type >= 0)
+			match = tag->type & q->type;
+
+		/* Remove tag from the results. tags are unowned so removing is easy */
+		if (!match)
+			g_queue_unlink(&res, node);
+	}
+
+	/* Convert GQueue to GPtrArray for sort, dedup and return */
+	i = 0;
+	ret = g_ptr_array_sized_new(g_queue_get_length(&res));
+	foreach_list(node, res.head)
+	{
+		tag = (TMTag *) node->data;
+		g_ptr_array_insert(ret, i++, tag);
+	}
+	g_list_free(res.head);
```

Comment:
A doubly-linked list is better when you insert in the middle or even beginning which this function does (arrays require the tail to be `memmove`'d. What makes you think direct `GPtrArray` + `tm_tags_prune()` is better?